go/vt/vttablet/tabletmanager/semisyncmonitor/monitor.go

```diff
@@ -0,0 +1,390 @@
+		return nil
+	}
+	log.Infof("Waiting for semi-sync to be unblocked")
+	// The primary is blocked. We need to wait for it to be unblocked
+	// or the context to expire.
+	ch := m.addWaiter()
+	select {
+	case <-ch:
+		log.Infof("Finished waiting for semi-sync to be unblocked")
+		return nil
+	case <-ctx.Done():
+		log.Infof("Error while waiting for semi-sync to be unblocked - %s", ctx.Err().Error())
+		return ctx.Err()
+	}
+}
+
+// stillBlocked returns true if the monitor should continue writing to the DB
+// because the monitor is still open, and the primary is still blocked.
+func (m *Monitor) stillBlocked() bool {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	return m.isOpen && m.isBlocked
+}
+
+// checkAndSetIsWriting checks if the monitor is already writing to the DB.
+// If it is not, then it sets the isWriting field and signals the caller.
+func (m *Monitor) checkAndSetIsWriting() bool {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if m.isWriting {
+		return false
+	}
+	m.isWriting = true
+	return true
+}
+
+// clearIsWriting clears the isWriting field.
+func (m *Monitor) clearIsWriting() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.isWriting = false
+}
+
+// startWrites starts writing to the DB.
+// It is re-entrant and will return if we are already writing.
+func (m *Monitor) startWrites() {
+	// If we are already writing, then we can just return.
+	if !m.checkAndSetIsWriting() {
+		return
+	}
+	// We defer the clear of the isWriting field.
+	defer m.clearIsWriting()
+
+	// Check if we need to continue writing or not.
+	for m.stillBlocked() {
+		// We do the writes in a go-routine because if the network disruption
+		// is somewhat long-lived, then the writes themselves can also block.
+		// By doing them in a go-routine we give the system more time to recover while
+		// exponentially backing off. We will not do more than maxWritesPermitted writes and once
+		// all maxWritesPermitted writes are blocked, we'll wait for VTOrc to run an ERS.
+		go m.write()
+		<-time.After(waitBetweenWrites)
+	}
+}
+
+// incrementWriteCount tries to increment the write count. It
+// also checks that the write count value should not exceed
+// the maximum value configured. It returns whether it was able
+// to increment the value or not.
+func (m *Monitor) incrementWriteCount() bool {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if m.inProgressWriteCount == maxWritesPermitted {
+		return false
+	}
+	m.inProgressWriteCount = m.inProgressWriteCount + 1
+	m.writesBlockedGauge.Set(int64(m.inProgressWriteCount))
+	return true
+}
+
+// AllWritesBlocked returns if maxWritesPermitted number of writes
+// are already outstanding.
+func (m *Monitor) AllWritesBlocked() bool {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	return m.isOpen && m.inProgressWriteCount == maxWritesPermitted
+}
+
+// decrementWriteCount decrements the write count.
+func (m *Monitor) decrementWriteCount() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.inProgressWriteCount = m.inProgressWriteCount - 1
```

Comment:
We can also use the decrement operator (--) here.

Reply:
Fixed